### PR TITLE
Uwuification update.

### LINKED ISSUE #812: Uwuification unsatisfactory.
## Description and reasoning

The current uwuification algorithm is very simple and frankly unsatisfactory, as there have been many recent advances in the field of uwuification I believe that Sir Lancebot would be greatly improved by the addition of a more feature-rich uwugorithm.

## Proposed Implementation

I personally have written a more advanced [uwuifier](https://github.com/gitautas/slowo) in python that does not depend on any external libraries and am willing to implement it into Sir Lancebot.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [ ] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by gitautas**
## Relevant Issues
Closes #812 

## Description
I ported over the implementation from my [uwuifier](https://github.com/gitautas/slowo) and removed the old command.

Old uwu:
image

New uwu: 
image

## Did you:
- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] Read all the comments in this template?
- [x] Ensure there is an issue open, or link relevant discord discussions?
- [x] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

**Comment by Xithrius** [comment]
It's finally time.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:15]: cute, but fuck no. Please use proper english in the code.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:45]: how about having it nonce

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:79]: I think it takes up a whole lot of vertical space for very little benefit when we put these regexes inbetween the helper functions instead of grouping them all together at the top. Yeah yeah, it has better locality, but it just makes this file so much more bloated.

I suggest we move all of these regexes and `subst` variables up below the constants, all together in a block. I don't think that will really make this harder to work with, but it'll make the file look better.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:96]: Can we maybe use full variable names instead of stuff like `subst`.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:99]: No reason why these many helper functions shouldn't have docstrings that explain what specifically they're used for, with an example.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:119]: kind of a pointless floating comment. Just put doctrings in stuff instead.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:139]: docstrings should be written in imperative mood.

This docstring needs to give examples.

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:141]: Newlines above block comments improve readability, and this comment isn't very clear. Convert it to what?

File: bot/exts/fun/uwu.py
Comment (lemonsaurus) [inline bot/exts/fun/uwu.py:146]: This comment also could use a newline before it, and I can't tell from the comment alone why you're doing this. Comments should explain the developers intentions.

I know you didn't write these comments, but you should follow the boyscout rulewhen interacting with our code; always leave the code better than you found it.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Can we please disable the emojis completely, it's so incredibly noisy and awful.

Other than that, I think this needs a little polish but the implementation itself looks okay.

**Comment by gitautas** [comment]
@lemonsaurus I'm pretty attached to the emojis, how about we turn down the weight and maybe get rid of some of the more noisy ones?

**Comment by gitautas** [comment]
@lemonsaurus I'm pretty attached to the emojis, how about we turn down the weight and maybe get rid of some of the more noisy ones?

> @lemonsaurus I'm pretty attached to the emojis, how about we turn down the weight and maybe get rid of some of the more noisy ones?

Here's the emojification based on punctuation and newlines and turned way down, what do you think?

image

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:55]: The `r''` actually stands for raw, it is mainly useful with backslashes. `r'\n'` becomes the actual string `'\n'` while `'\n'` in the code becomes a new-line character. It is pretty common to think `r` is for regex but this is not correct.

The reason I say this is because you double-backslash in the subtitute strings, and always use raw strings in the regex compilation.

*Side note: these should be all-caps for CONSTANTS*.

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:110]: Can't these be placed in the cog itself?

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:122]:
```suggestion
        Echo an uwuified version of the passed text.
```

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:137]: What is this if-statement for?

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py:26]: Lookup table.

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py:55]: That's good information, will update when I can.

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py]: Done.

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py]: Done.

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py:26]: Anything wrong with this?

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:67]: Might as well use `dict.items()` like this:
```suggestion
        for word, replacement in WORD_REPLACE_HASH.items():
            input_string = input_string.replace(word, replacement)
```

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:71]: It doesn't replace all characters so we can clarify that like this:
```suggestion
        """Replace certain characters with 'w'."""
```
Do feel free to improve it in other ways.

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:78]: Is this typing correct? A callable doesn't have a `string` attribute?

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:83]:
```suggestion
            return f"{match_string}-{match_string[-1]}"  # Stutter the last character
```

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:87]:
```
        """Nyaifies a string by doing XYZ."""
```
Can you just add some more details here?

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:98]: Are you looking for [`random.choice`](https://docs.python.org/3/library/random.html#random.choice)?

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [reply bot/exts/fun/uwu.py:78]:
```suggestion
    def _stutter_replace(self, match: re.Match, strength: float = 0.0) -> str:
```

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [inline bot/exts/fun/uwu.py:94]:
```suggestion
    def _emoji_replace(self, match: re.Match, strength: float = 0.0) -> str:
```

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [inline bot/exts/fun/uwu.py:96]: This can be simplified.
```suggestion
        match_string = match.group()
```

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [inline bot/exts/fun/uwu.py:80]:
```suggestion
        match_string = match.group()
```

**Review by gitautas — COMMENTED** [review]
Commiting suggestions.

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [reply bot/exts/fun/uwu.py:26]: I wouldn't really say it is a big issue no. A table usually refers to a dictionary though and `LUT` may not be the most common acronym you can find so-to-speak (at least from my experience)..

If you would like to keep it as `EMOJI_LUT` then that's fine but changing it to `EMOJIS` could make it more clear. Especially considering you're using it with `random.choice()` later on.

**Comment by Xithrius** [comment]
Greetings.

Please see [this page](https://www.pythondiscord.com/pages/guides/pydis-guides/contributing/contributing-guidelines/supplemental-information/#writing-good-commit-messages) on making commits with great messages, to increase users' knowledge of the update history.

Thanks!

File: bot/exts/fun/uwu.py
Comment (onerandomusername) [inline bot/exts/fun/uwu.py:52]: Curious, why isn't this compiled too?

File: bot/exts/fun/uwu.py
Comment (onerandomusername) [inline bot/exts/fun/uwu.py:55]: Same here

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [reply bot/exts/fun/uwu.py:52]: The `repl` argument of [`re.sub()`](https://docs.python.org/3/library/re.html#re.sub) requires a string, not a pattern object.

**Comment by Xithrius** [comment]
@gitautas if you're not able to work on this PR anymore, would you like it to go up for grabs?

**Comment by shenanigansd** [comment]
Per discussion on Discord, I will work on this PR on behalf of @gitautas.
https://discord.com/channels/267624335836053506/412357430186344448/946087902058774568

**Comment by gitautas** [comment]
Heck yeah!

File: bot/exts/fun/uwu.py
Comment (shenanigansd) [reply bot/exts/fun/uwu.py:51]: https://github.com/python/typeshed/blob/master/stdlib/typing.pyi#L656
The type hint for the `repl` argument is `AnyStr`.

https://github.com/python/cpython/blob/main/Lib/re.py#L277-L281
But the cpython source shows that a compiled pattern is allowed if there are no flags.

Is the compiled pattern okay, or is it required to respect the type hint, and to leave the pattern as a string?

File: bot/exts/fun/uwu.py
Comment (shenanigansd) [reply bot/exts/fun/uwu.py]: Changed the name to `EMOJIS`.

File: bot/exts/fun/uwu.py
Comment (shenanigansd) [reply bot/exts/fun/uwu.py]: Update this to use [`random.choice`](https://docs.python.org/3/library/random.html#random.choice).

File: bot/exts/fun/uwu.py
Comment (shenanigansd) [reply bot/exts/fun/uwu.py]: Removed unused import.

File: bot/exts/fun/uwu.py
Comment (shenanigansd) [reply bot/exts/fun/uwu.py:51]: Never mind. It can't be compiled ahead of time anyways because the backreference doesn't exist yet.

File: bot/exts/fun/uwu.py
Comment (Bluenix2) [inline bot/exts/fun/uwu.py:121]: We should change this to use `get_cog()`.
 ```suggestion
         if (fun_cog := ctx.bot.get_cog("Fun")):
             text, embed = await fun_cog._get_text_and_embed(ctx, text)

            # Grabs the text from the embed for uwuification.
            if embed is not None:
                embed = fun_cog._convert_embed(self._uwuify, embed)
        else:
             embed = None
```

Comment (gitautas) [reply bot/exts/fun/uwu.py]: Linting fails for the suggestion

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py]: Thanks for the update!

File: bot/exts/fun/uwu.py
Comment (gitautas) [reply bot/exts/fun/uwu.py]: Done.

**Comment by gitautas** [comment]
Alright I think that's everything, now we need a review.

**Review by Bluenix2 — COMMENTED** [review]
I just have one comment about an issue I ran into trying to use the embed feature - other than that this looks great!

File: bot/exts/fun/uwu.py
Comment (ToxicKidz) [reply bot/exts/fun/uwu.py:121]: Doesn't [`Fun._get_text_and_embed()`](https://github.com/python-discord/sir-lancebot/blob/76819000fdad3820397b67ceba5a1d4192dc50b6/bot/exts/fun/fun.py#L153-L177) already do this?

Comment (ToxicKidz) [reply bot/exts/fun/uwu.py:121]: This might seem unreliable, since embeds come *after* the message is sent.

Comment (Bluenix2) [reply bot/exts/fun/uwu.py:121]: Hmm, you're right. I didn't consider that.

**Review by Bluenix2 — APPROVED** [review]
After discussing with @ToxicKidz, the one comment I had has been resolved. Thank you for this feature, it's a big improvement!

**Comment by gitautas** [comment]
@lemonsaurus Can you look at the new emoji value and see if that satisfies your high standards for uwuification? ;)
Your change request is blocking the PR.

**Comment by janine9vn** [comment]
@gitautas Just an FYI since our repos operate a bit differently, the change request isn't blocking the PR. You do need a core dev to review and approve for ~~our overlord~~ policy bot to be satisfied.

**Comment by gitautas** [comment]
> @gitautas Just an FYI since our repos operate a bit differently, the change request isn't blocking the PR. You do need a core dev to review and approve for ~our overlord~ policy bot to be satisfied.

Ohh, I see! Thanks for the information!

**Pull request merged**
